tensor2tensor/trax/layers/initializers.py

```diff
@@ -30,21 +41,70 @@ def Init(shape, rng):
+  """Initializer capable of adapting its scale to the shape of weights tensors."""
+  if scale <= 0.:
+    raise ValueError('scale must be positive float, {} given'.format(scale))
+  if mode not in {'fan_in', 'fan_out', 'fan_avg'}:
+    raise ValueError('Invalid mode argument:, {}, must be either fan_in, fan_out or fan_avg'.format(mode))
+
   def Init(shape, rng):
-    fan_in, fan_out = shape[in_dim], shape[out_dim]
-    std = backend.numpy.sqrt(2.0 / (fan_in + fan_out))
-    a = backend.numpy.sqrt(3.0) * std
-    return backend.random.uniform(rng, shape, minval=-a, maxval=a)
-  return Init
+    fan_in, fan_out = _get_fans(shape, out_dim, in_dim)
+    gain = scale
+    if mode == 'fan_in':
+      gain /= fan_in
+    elif mode == 'fan_out':
+      gain /= fan_out
+    elif mode == 'fan_avg':
+      gain /= (fan_in + fan_out) / 2
+    if distribution == 'truncated_normal':
+      # constant from scipy.stats.truncnorm.std(a=-2, b=2, loc=0., scale=1.)
+      stddev = backend.numpy.sqrt(gain) / .87962566103423978
+      return (backend.random.truncated_normal(rng, -2, 2, shape) * stddev).astype('float32')
+    elif distribution == 'normal':
+      return (backend.random.normal(rng, shape, dtype) * backend.numpy.sqrt(gain)).astype('float32')
+    elif distribution == 'uniform':
+      lim = backend.numpy.sqrt(3. * gain)
+      return backend.random.uniform(rng, shape, dtype, minval=-lim, maxval=lim).astype('float32')
+    else:
+      raise ValueError('invalid distribution for variance scaling Initializer')
+    return Init
+
+
+def GlorotNormalInitializer(out_dim=0, in_dim=1, scale=1.):
+  """An initializer function for random Glorot-scaled coefficients."""
+  return VarianceScalingInitializer(out_dim, in_dim, scale, 'fan_avg', 'truncated_normal')
+
+
+def GlorotUniformInitializer(out_dim=0, in_dim=1, scale=1.):
+  """An initializer function for random uniform Glorot-scaled coefficients."""
+  return VarianceScalingInitializer(out_dim, in_dim, scale, 'fan_avg', 'uniform')    
+
+
+def LeCunNormalInitializer(out_dim=0, in_dim=1, scale=1.):
+  """An initializer function for random LeCun-scaled coefficients."""
+  return VarianceScalingInitializer(out_dim, in_dim, scale, 'fan_in', 'truncated_normal')
+
+
+def LeCunUniformInitializer(out_dim=0, in_dim=1):
+  """An initializer function for random uniform LeCun-scaled coefficients."""
+  return VarianceScalingInitializer(out_dim, in_dim, scale, 'fan_in', 'uniform')    
+
+
+def KaimingNormalInitializer(out_dim=0, in_dim=1, param=0.):
+  """An initializer function for random Kaiming-scaled coefficients."""
+  return VarianceScalingInitializer(out_dim, in_dim, 2.0 / backend.np.sqrt(1 + param**2),
+                                    'fan_in', 'truncated_normal')
+
+
+def KaimingUniformInitializer(out_dim=0, in_dim=1, param=0.):
+  """An initializer function for random uniform Kaiming-scaled coefficients."""
+  return VarianceScalingInitializer((out_dim, in_dim, 2.0 / backend.np.sqrt(1 + param**2),
```

Comment:
This has double brackets, so it's not correct python?

Reply:
Yes, weird how those tests were passing, thanks for catching it!